### ISSUE TITLE
Feature request: Support for S3 Object Locks

### ISSUE THREAD

**Issue by gwvandesteeg** [terraform-providers/terraform-provider-aws#6634]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support for the S3 Object Lock on the S3 bucket itself, and the S3 objects themselves as per the documentation.
* S3 Object lock requires versioning to be enabled
* S3 Object locks cannot be added to an existing bucket at this time

### New or Affected Resource(s)

* aws_s3_bucket
Additionally would need to add the object specific locks depending on the API

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
resource "aws_s3_bucket" "lockme" {
  bucket = "some-bucket-name"
  versioning {
    enabled = true
    object_lock = true
  }
}
```

### References

* https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock.html

**Comment by bflad** [comment]
Support for S3 object lock configuration in the `aws_s3_bucket` resource has been merged and will release with version 1.56.0 of the Terraform AWS provider, likely in the next few hours.

**Comment by bflad** [comment]
This has been released in [version 1.56.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1560-january-16-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
